### PR TITLE
Hotfix/#1169 extra lazy one to many should not delete referenced entities

### PR DESCRIPTION
**Pull request by Ocramius**
This is an alternate fix for #1169 (DDC-3343 http://www.doctrine-project.org/jira/browse/DDC-3343)

### CODE REVIEW

**Comment by stof** [comment]
looks good to me

**Comment by Ocramius** [comment]
@stof I'll provide a 2.4 PR as well in a few.

File: tests/Doctrine/Tests/ORM/Functional/ValueConversionType/OneToManyExtraLazyTest.php [hunk 1]
```diff
@@ -103,4 +105,101 @@ public function testThatASliceOfTheExtraLazyCollectionIsLoaded()
 
         $this->assertCount(2, $inversed->associatedEntities->slice(0, 2));
     }
+
+    /**
+     * @group DDC-3343
+     */
+    public function testRemovesManagedElementFromOneToManyExtraLazyCollection()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/ValueConversionType/OneToManyExtraLazyTest.php]: These tests are actually on the wrong file.

File: lib/Doctrine/ORM/Cache/Persister/Collection/AbstractCollectionPersister.php [hunk 2]
```diff
@@ -264,4 +263,42 @@ public function loadCriteria(PersistentCollection $collection, Criteria $criteri
     {
         return $this->persister->loadCriteria($collection, $criteria);
     }
+
+    /**
+     * Clears cache entries related to the current collection
+     *
+     * @param PersistentCollection $collection
+     */
+    protected function evictCollectionCache(PersistentCollection $collection)
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Cache/Persister/Collection/AbstractCollectionPersister.php]: I figured that none of the `EXTRA_LAZY` stuff was actually updating the collection-related caches. This change fixes that

File: tests/Doctrine/Tests/ORM/Cache/Persister/Collection/AbstractCollectionPersisterTest.php [hunk 3]
```diff
@@ -240,22 +240,6 @@ public function testInvokeRemoveElement()
         $this->assertFalse($persister->removeElement($collection, $element));
     }
 
-    public function testInvokeRemoveKey()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Cache/Persister/Collection/AbstractCollectionPersisterTest.php]: Method was dead/unused

File: lib/Doctrine/ORM/Cache/Persister/Collection/AbstractCollectionPersister.php [hunk 4]
```diff
@@ -238,15 +239,13 @@ public function get(PersistentCollection $collection, $index)
...
+        if ($persisterResult = $this->persister->removeElement($collection, $element)) {
+            $this->evictCollectionCache($collection);
+            $this->evictElementCache($this->sourceEntity->rootEntityName, $collection->getOwner());
+            $this->evictElementCache($this->targetEntity->rootEntityName, $element);
+        }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function removeKey(PersistentCollection $collection, $key)
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Cache/Persister/Collection/AbstractCollectionPersister.php]: Method was dead/unused

File: lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php [hunk 5]
```diff
@@ -163,7 +164,20 @@ public function removeElement(PersistentCollection $collection, $element)
         $mapping   = $collection->getMapping();
         $persister = $this->uow->getEntityPersister($mapping['targetEntity']);
 
-        return $persister->delete($element);
```
Comment (beberlei) [inline lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: I think this was intended to be used only be used with `orphanRemoval=true`, not always like it is now.

File: lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php [hunk 6]
```diff
@@ -163,7 +164,20 @@ public function removeElement(PersistentCollection $collection, $element)
...
+        $targetMetadata = $this->em->getClassMetadata($mapping['targetEntity']);
+
+        if ($element instanceof Proxy && ! $element->__isInitialized()) {
+            $element->__load();
+        }
+
+        // clearing owning side value
+        $targetMetadata->reflFields[$mapping['mappedBy']]->setValue($element, null);
```
Comment (beberlei) [inline lib/Doctrine/ORM/Persisters/Collection/OneToManyPersister.php]: This allows the "inverse" side to change the "owning-side", which is exactly what we didn't want to do.

**Comment by Ocramius** [comment]
To recap, I need to fix following:

 - when using `orphan-removal`, this triggers a `DELETE` operation on the inverse side
 - when **not** using `orphan-removal`, this becomes no-op.

**Comment by Ocramius** [comment]
Provided required hotfixes via #1288

**Pull request merged**